File: server/collector-engine/strategia_collector/connectors/rss_connector.py
import logging
import feedparser
from urllib.parse import urlparse

from strategia_collector.connectors.connector_base import BaseConnector, CollectedData

logger = logging.getLogger(__name__)


class RssConnector(BaseConnector):
    TIMEOUT = 10

    async def fetch(self, url: str) -> list[CollectedData]:
        logger.info(f"Fetching RSS feed: {url}")

        try:
            feed = feedparser.parse(url)
            items = []

            for entry in feed.entries[:20]:
                content = entry.get("content", [{}])[0].get("value", "") if entry.get("content") else ""
                content = content or entry.get("description", "") or entry.get("summary", "")

                pub_date = None
                if entry.get("published"):
                    pub_date = self._parse_date(entry.published)
                elif entry.get("updated"):
                    pub_date = self._parse_date(entry.updated)

                items.append(CollectedData(
                    source_url=entry.get("link", url),
                    source_type="RSS",
                    title=entry.get("title", ""),
                    description=content[:200],
                    content=content,
                    content_raw=content,
                    published_at=pub_date,
                ))

            feed_title = feed.feed.get("title", urlparse(url).netloc)
            logger.info(f"RSS: fetched {len(items)} items from {feed_title}")
            return items

        except Exception as e:
            logger.error(f"RSS fetch failed for {url}: {e}")
            raise

    @staticmethod
    def _parse_date(date_str: str):
        from datetime import datetime
        try:
            return datetime.fromisoformat(date_str.replace("Z", "+00:00"))
        except Exception:
            try:
                return datetime.strptime(date_str, "%a, %d %b %Y %H:%M:%S %z")
            except Exception:
                return None

File: server/src/modules/collection-engine/connectors/rss.connector.ts
/**
 * FICHIER: rss.connector.ts
 *
 * RÔLE: Connecteur RSS qui parse les flux RSS/Atom et extrait les articles (20 max par flux).
 *
 * RESPONSABILITÉS:
 * - Parser les flux RSS/Atom avec rss-parser
 * - Extraire titre, description, contenu brut, date de publication
 * - Limiter à 20 items par flux
 *
 * FLUX:
 * - CollectionEngineService → RssConnectorService.fetch(url) → CollectedData[]
 *
 * EXEMPLE: Un flux RSS d'actualités est parsé → 20 articles extraits avec titre et contenu.
 */
import { Injectable, Logger } from '@nestjs/common';
import * as RssParser from 'rss-parser';
import { IConnector, CollectedData } from './connector.interface';

@Injectable()
export class RssConnectorService implements IConnector {
  private readonly logger = new Logger(RssConnectorService.name);
  private readonly MAX_ITEMS = 20;
  private parser: RssParser;

  constructor() {
    this.parser = new RssParser({
      timeout: 30000,
      headers: {
        'User-Agent': 'Mozilla/5.0 (compatible; StrategIA-Collector/1.0)',
        Accept: 'application/rss+xml, application/atom+xml, application/xml',
      },
    });
  }

  async fetch(url: string): Promise<CollectedData[]> {
    try {
      const feed = await this.parser.parseURL(url);
      this.logger.log(`RSS: ${feed.items.length} items from ${url}`);
      return feed.items.slice(0, this.MAX_ITEMS).map(item => ({
        sourceUrl: item.link || url,
        sourceType: 'RSS',
        title: item.title || '',
        description: (item.contentSnippet || item.content || '').substring(0, 300),
        contentRaw: item.content || item.contentSnippet || '',
        publishedAt: item.pubDate ? new Date(item.pubDate) : undefined,
      }));
    } catch (error: any) {
      this.logger.error(`RSS fetch failed for ${url}: ${error.message}`);
      return [];
    }
  }
}
